### DISCUSSION TITLE
ISL28022 power monitor in addition to the INA219 power monitor.

### DISCUSSION THREAD

**Discussion by mhardy6351** [arendst/Tasmota#14356] [Ideas]
I have a need to monitor an eBike battery with voltage up to 60 volts DC.

The INA219 supports up to 32 volts.

I found an I2C bus supports ISL28022 power monitor that is discovered with the I2Cscan and displays as an INA219 device.

That's great for any voltage below 32 volts. But top out at 32 volts when I attach the battery and charger that outputs up to 54.6 volts.

Is there a way to set the upper voltage limit?

I have added a total of 3 of these to one D1 mini and everything works below 32 volts. I have not tried to test the current and power values since the voltage is off at the moment.

https://curiouselectric.co.uk/collections/monitoring/products/dc-power-sensor-i2c

Thanks in advance.

Mike

**Comment by mhardy6351** [comment]
There is no voltage applied at the moment, 

05:13:06.520 CMD: I2cscan
05:13:06.541 MQT: DC_volts/RESULT = {"I2CScan":"Device(s) found at 0x3f 0x40"}
05:13:18.415 MQT: DC_volts/SENSOR = {"Time":"2022-01-10T05:13:18","INA219":{"Id":40,"Voltage":1,"Current":-0.000,"Power":-0}}

**Reply by mhardy6351** [reply]
I have flashed the latest development sensors.bin to the D1 Mini

**Reply by barbudor** [reply]
Hi
In a normal situation I would have said that you are unlucky as the ISL28022 has been detected as an INA219 because they share the same 0x40 address but as it's a different chip it would needs to have it's own driver (this kind of problem occurs a lot in I2C as very different chips can use the same address see table in [I2CDEVICES](https://tasmota.github.io/docs/I2CDEVICES/).

However looking quickly at the datasheet of the ISL28022 it looks like it is a extension of the INA219 with all 5 first registers being very simular. So it should be possible to make the existing INA219 driver to support also the ISL28022.

I will look in more details tonight the difference between the 2 chips and see if we can extend the INA219 driver or if a totally new driver is needed. But as the chips are very close, I should be able to do this without having access to the chip itself.

The driver supports up to 4 INA219 chips but the Sensor13 command will impact all 4 (so you need to use the same current shunt on all of them).

**Reply by mhardy6351** [reply]
Thank you for taking the time to look into this.

With modern devices like eBikes, scooters and one wheels, the prevalence of batteries and power sources above 30 volts are quite common.

There is sample code for Arduino on the device GitHub repository if this may be of any help.

https://github.com/curiouselectric/ISL28022_breakout

I have no issue with the same valued shunt being used on the I2C bus. If different values is required, then I will use multiple controllers. 

The device has 0.005 ohm shunt supplied on the board.

Please let me know if I can be of any other help

**Reply by barbudor** [reply]
> I have no issue with the same valued shunt being used on the I2C bus. If different values is required, then I will use multiple controllers.
> The device has 0.005 ohm shunt supplied on the board.

That 0.005 ohms is the shunt I was talking about for measuring current. Not related to I2C pull-ups.

I will look tonight

**Reply by mhardy6351** [reply]
Do you think this is supportable?

Plan B is to get an Arduino to use these sensors.

**Reply by barbudor** [reply]
Apologies, I had to address some other topics in the mean time but I will work on this and provide you with a test version before the end of the weekend. Either as an extension of existing INA219 or as a new driver. 
Thank you for your patience.

**Reply by mhardy6351** [reply]
Thank you and I understand priorities.

I look forward to testing.

**Reply by barbudor** [reply]
@mhardy6351 Please test the following binary of tasmota-sensors.
It's the standard tasmota-sensors but INA219 driver has been expanded to be interoperable with ISL28022 in 60V range.
It is still working fine with my INA219 and should transparently work with your ISL28022
One thing I didn't tested is more than 1 device.

Please torture that as much as possible across the range, and if your system supports it with negative currents too.

The Web and sensor message still display INA219 as the way I've updated the driver, it doesn't need to know if it is an INA219 or an ISL28022 as the registers are some much interoperable.
If you feel that it would be preferable to have the proper sensor name, I can make a #define override option just for the name but that will requires you build your own firmware instead of using pre-built tasmota-sensors.bin

I have not implemented anything related to the alarm/threshold features that are ISL28022 specific.
If you have a use case for it, I could implemented them also but that would definitively requires a #define override option in that case.

Looking forward your (positive) feedback.

[tasmota-isl28022v1.zip](https://github.com/arendst/Tasmota/files/7866447/tasmota-isl28022v1.zip)

Forgot to mention that for a 5 milliohms shunt you will need to enter `Sensor13 50`  = 5 * 10^0 milliohms
See the command [Sensor13](https://tasmota.github.io/docs/Commands/#sensors) for more details.

**Reply by mhardy6351** [reply]
Thank you for this.

The firmware loaded correctly and the device rebooted.

First observation, on the web portal, the old firmware would display up to 32 volts DC.

This version stops at 27 volts DC.

I have only tested one 54.6 volt charger attached to the device on ID 40.
I observe minor current fluctuations in current, but the battery is fully charged to 54.6 volts. I will be able to tell more when the voltage is addresses and I test with a second and third power source.

![image](https://user-images.githubusercontent.com/83194086/149554053-f96d8509-5efa-4cd0-a470-39312431ed03.png)

CMD: I2cscan

17:39:06.223 MQT: DC_volts/RESULT = {"I2CScan":"Device(s) found at 0x3f 0x40 0x44 0x45"}

17:39:33.656 MQT: DC_volts/SENSOR = {"Time":"2022-01-14T17:39:33","INA219-1":{"Id":40,"Voltage":27,"Current":0.000,"Power":0},"INA219-2":{"Id":44,"Voltage":-0,"Current":-0.000,"Power":0},"INA219-3":{"Id":45,"Voltage":-0,"Current":-0.000,"Power":0}}
17:40:03.700 MQT: DC_volts/SENSOR = {"Time":"2022-01-14T17:40:03","INA219-1":{"Id":40,"Voltage":27,"Current":0.000,"Power":0},"INA219-2":{"Id":44,"Voltage":-0,"Current":-0.000,"Power":0},"INA219-3":{"Id":45,"Voltage":-0,"Current":-0.000,"Power":0}}

17:40:33.688 MQT: DC_volts/SENSOR = {"Time":"2022-01-14T17:40:33","INA219-1":
17:41:03.628 MQT: DC_volts/SENSOR = {"Time":"2022-01-14T17:41:03","INA219-1":{"Id":40,"Voltage":27,"Current":0.000,"Power":0},"INA219-2":{"Id":44,"Voltage":-0,"Current":-0.000,"Power":0},"INA219-3":{"Id":45,"Voltage":-0,"Current":-0.001,"Power":0}}


Please let me know if you need any other data sent.

**Reply by barbudor** [reply]
So it looks you say the voltage displayed is half of what was expected.
Need to read again the datasheet

**Reply by mhardy6351** [reply]
Yes, voltres 3 displays 27.148 volts.

So yes 1/2 appears correct.

**Reply by barbudor** [reply]
Ok, reading better the datasheet, I finally need to detect if the chip is INA219 or ISL28022 for reading the Bus Voltage register. My head was not too fresh yesterday evening.
Please try this version.
Now the Web and SENSOR should say ISL28022 instead of INA219 which would confirm I'm detecting the chip correctly.
If not, I would need a copy of the ~10 first lines after boot with Weblog 3.
It should looks like those :
```
0:00:00.056 QPC: Count 1
00:00:00.062 ROT: Mode 1
00:00:00.065 Project tasmota - nodemcu Version 2022.01.2(sensors)-2_7_4_9(2022-01-14T19:34:23)
00:00:00.067 Ina219SetCalibration: mode=1
00:00:00.067 Ina219SetCalibration: cur_mul=0.01000
00:00:00.069 Ina219Ident: IntrReg=0x3E67 (1)
00:00:00.069 I2C: INA219 found at 0x40
00:00:00.080 Ina219GetBusVoltage_V: BusReg = 0x0333 (819)
00:00:00.081 Ina219GetShuntVoltage_mV: ShReg = 0x0147 (327)
00:00:00.082 Ina219Read: bV=3.27600V, sV=3.27000mV
00:00:00.082 Ina219Read: V=3.27927V, I=0.03270mA
```
[tasmota-isl28022v2.zip](https://github.com/arendst/Tasmota/files/7872528/tasmota-isl28022v2.zip)

**Reply by mhardy6351** [reply]
It appears correct at this time.



00:00:00.067 Project tasmota - DC_volts Version 2022.01.2(sensors)-2_7_4_9(2022-01-14T19:34:23)
00:00:00.113 I2C: ISL28022 found at 0x40
00:00:00.116 I2C: ISL28022 found at 0x41
00:00:00.120 I2C: ISL28022 found at 0x44
00:00:00.515 WIF: Connecting to AP1 Hardy Channel 11 BSSId A6:40:A0:61:48:6E in mode 11n as DC-volts-5444...
00:00:03.755 WIF: Connected
00:00:04.008 HTP: Web server active on DC-volts-5444 with IP address 192.168.1.16
21:00:10.012 MQT: Attempting connection...
21:00:10.227 MQT: Connected
21:00:10.232 MQT: DC_volts/LWT = Online (retained)
21:00:10.235 MQT: DC_volts/cmnd/POWER = 
21:00:10.244 MQT: DC_volts/INFO1 = {"Info1":{"Module":"Generic","Version":"2022.01.2(sensors)","FallbackTopic":"cmnd/DVES_9B5544_fb/","GroupTopic":"DC_volts/cmnd/"}}

**Reply by mhardy6351** [reply]
21:04:44.588 MQT: DC_volts/RESULT = {"WebLog":2}
21:04:47.665 MQT: DC_volts/STATE = {"Time":"2022-01-14T21:04:47","Uptime":"0T00:04:44","UptimeSec":284,"Heap":21,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":31,"MqttCount":1,"Wifi":{"AP":1,"SSId":"Hardy","BSSId":"A6:40:A0:61:48:6E","Channel":11,"Mode":"11n","RSSI":72,"Signal":-64,"LinkCount":1,"Downtime":"0T00:00:05"}}
21:04:47.694 MQT: DC_volts/SENSOR = {"Time":"2022-01-14T21:04:47","ISL28022-1":{"Id":40,"Voltage":54.296,"Current":-0.032,"Power":-2},"ISL28022-2":{"Id":41,"Voltage":0.000,"Current":0.004,"Power":0},"ISL28022-3":{"Id":44,"Voltage":0.000,"Current":0.000,"Power":0}}
21:05:17.790 MQT: DC_volts/STATE = {"Time":"2022-01-14T21:05:17","Uptime":"0T00:05:14","UptimeSec":314,"Heap":21,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"Wifi":{"AP":1,"SSId":"Hardy","BSSId":"A6:40:A0:61:48:6E","Channel":11,"Mode":"11n","RSSI":66,"Signal":-67,"LinkCount":1,"Downtime":"0T00:00:05"}}
21:05:17.821 MQT: DC_volts/SENSOR = {"Time":"2022-01-14T21:05:17","ISL28022-1":{"Id":40,"Voltage":54.300,"Current":-0.032,"Power":-2},"ISL28022-2":{"Id":41,"Voltage":0.000,"Current":0.002,"Power":0},"ISL28022-3":{"Id":44,"Voltage":0.000,"Current":0.002,"Power":0}}
21:05:47.699 MQT: DC_volts/STATE = {"Time":"2022-01-14T21:05:47","Uptime":"0T00:05:44","UptimeSec":344,"Heap":21,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"Wifi":{"AP":1,"SSId":"Hardy","BSSId":"A6:40:A0:61:48:6E","Channel":11,"Mode":"11n","RSSI":58,"Signal":-71,"LinkCount":1,"Downtime":"0T00:00:05"}}
21:05:47.728 MQT: DC_volts/SENSOR = {"Time":"2022-01-14T21:05:47","ISL28022-1":{"Id":40,"Voltage":0.028,"Current":0.002,"Power":0},"ISL28022-2":{"Id":41,"Voltage":0.000,"Current":0.002,"Power":0},"ISL28022-3":{"Id":44,"Voltage":0.000,"Current":0.002,"Power":0}}
21:06:17.700 MQT: DC_volts/STATE = {"Time":"2022-01-14T21:06:17","Uptime":"0T00:06:14","UptimeSec":374,"Heap":21,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"Wifi":{"AP":1,"SSId":"Hardy","BSSId":"A6:40:A0:61:48:6E","Channel":11,"Mode":"11n","RSSI":60,"Signal":-70,"LinkCount":1,"Downtime":"0T00:00:05"}}
21:06:17.729 MQT: DC_volts/SENSOR = {"Time":"2022-01-14T21:06:17","ISL28022-1":{"Id":40,"Voltage":54.304,"Current":-0.032,"Power":-2},"ISL28022-2":{"Id":41,"Voltage":0.000,"Current":0.004,"Power":0},"ISL28022-3":{"Id":44,"Voltage":0.000,"Current":0.000,"Power":0}}

**Reply by barbudor** [reply]
Cool
Please also test Current
When you will confirm it is ok, I will push that to Tasmota development version.

**Reply by mhardy6351** [reply]
Thank you again for your adding support. Testing the voltage and current appears to be within range of my other voltage and current meter. 

I haven't had the opportunity to test negative current or fully test all 3 devices at once with voltage applied but I will do that today.

At this time, I feel you should be able to push the code to the master build. 

I'm not sure if I will need the alarms and other features of the isl20880. I will read the data sheet and see if it's a feature needed direct;y in Tasmota, or if it's something I just implement outside in node red.

Thank you again.

**Reply by barbudor** [reply]
Thanks for the confirmation, I will push the code tonight. Depending on Theo's availability to merge, that should be in the development version within a couple of days.

Regarding alarms, indeed you can still do in node-red (at Teleperiod), or if fast response is required with rules on Tasmota

**Reply by barbudor** [reply]
PR [#14558](https://github.com/arendst/Tasmota/pull/14458)

**Reply by barbudor** [reply]
PR merged

**Reply by Tino4** [reply]
Hi guys , i've noticed that there is a wrong calculation of the current and bus voltage from a ISL28022. A ISL28022 uses a measurement resistor of 5 milliohms instead of 100milliohm(ina219). And i believe that ISL28022 uses a part of the ina219 calculation code.
Best regards, Tino

**Reply by barbudor** [reply]
@Tino4 
The ISL28022 doesn't have a whatever milliohm resistor.
Neither the INA219 nor the ISL28022 have any built-in resistor at all.
For both chips, it is up to the hardware designer to use the appropriate shunt resistor to achieve the expected current range.
While there are many standard INA219 modules with 0.1 ohms available on the market, nothing prevent you to use the INA with whatever shunt you want. AFAIK there is no standard ISL28022 module.

The point is YOU have a ISL28022 setup with a 5 milliohm resistor and you should use [`Sensor13 50`](https://tasmota.github.io/docs/Commands/#sensor13) command to configure the shunt resistor.
By the way, same command for either INA219 and ISL28022. The driver takes care of the difference between the 2 chips.
